src/mapping/parser/mod.rs

```diff
@@ -195,151 +202,132 @@ fn query_arithmetic_from_pair(pair: Pair<Rule>) -> Result<Box<dyn query::Functio
-        }
+fn function_arguments_from_pairs(
+    pairs: Pairs<Rule>,
+    signature: &FunctionSignature,
+) -> Result<ArgumentList> {
+    let mut arguments = ArgumentList::new();
+
+    // Keeps track of positional argument indices.
+    //
+    // Used to map a positional argument to its keyword. Keyword arguments can
+    // be used in any order, and don't count towards the index of positional
+    // arguments.
+    let mut index = 0;
+
+    pairs
+        .map(|pair| pair.into_inner().next().unwrap())
+        .map(|pair| match pair.as_rule() {
+            Rule::positional_item => {
+                index += 1;
+                positional_item_from_pair(pair, &mut arguments, index - 1, signature)
+            }
+            Rule::keyword_item => keyword_item_from_pair(pair, &mut arguments, signature),
+            _ => unexpected_parser_sytax!(pair),
+        })
+        .collect::<Result<()>>()?;
 
-        _ => unreachable!("parser should not allow other query_function child rules here"),
+    // check invalid arity
+    if arguments.len() > signature.parameters().len() {
+        return Err(format!(
+            "invalid number of function arguments (got {}, expected {}) for function '{}'",
+            arguments.len(),
+            signature.parameters().len(),
+            signature.as_str(),
+        ));
     }
+
+    // check missing required arguments
+    signature
+        .parameters()
+        .iter()
+        .filter(|p| p.required)
+        .filter(|p| !arguments.keywords().contains(&p.keyword))
+        .map(|p| {
+            Err(format!(
+                "required argument '{}' missing for function '{}'",
+                p.keyword,
+                signature.as_str()
+            ))
+        })
+        .collect::<Result<_>>()?;
+
+    // check unknown argument keywords
+    arguments
+        .keywords()
+        .iter()
+        .filter(|k| !signature.parameters().iter().any(|p| &p.keyword == *k))
+        .map(|k| {
+            Err(format!(
+                "unknown argument keyword '{}' for function '{}'",
+                k,
+                signature.as_str()
+            ))
+        })
+        .collect::<Result<_>>()?;
+
+    Ok(arguments)
+}
+
+fn positional_item_from_pair(
+    pair: Pair<Rule>,
+    list: &mut ArgumentList,
+    index: usize,
+    signature: &FunctionSignature,
+) -> Result<()> {
+    let resolver = query_arithmetic_from_pair(pair.into_inner().next().ok_or(TOKEN_ERR)?)?;
+
+    let parameter = signature.parameters().get(index).cloned().ok_or(format!(
+        "unknown positional argument '{}' for function: '{}'",
+        index,
+        signature.as_str()
+    ))?;
+
+    let keyword = parameter.keyword.to_owned();
+    let argument = Argument::new(resolver, parameter);
+
+    // TODO(jean): I think we can get rid of the `ArgumentList` wrapper, as we
+    // always apply a keyword to an argument, even if a positional argument was
+    // given.
+    list.push(argument, Some(keyword));
```

Comment:
Will tidy this up before I merge. The `ArgumentList` type will remain, as it provides convenient functions `required` and `optional` that are used in `functions.rs`, but storing the keywords in that list has become pointless.